### DISCUSSION TITLE
PWM Control of Constant Current RGBW LED Drivers

### DISCUSSION THREAD

**Discussion by ArrEssJay** [arendst/Tasmota#15208] [Q&A] [answered]
I'm trying to achieve acceptable brightness control of an RGBW lamp (Mi Bedside Lamp) which uses PWM to servo 4x [RT8471](https://datasheet.lcsc.com/lcsc/1806070736_Richtek-Tech-RT8471GJ5_C105529.pdf) constant current LED drivers. 

The PWM duty cycle/current curve appears linear at 500Hz. I've set `PwmFrequency` accordingly.

<img width="366" alt="Screen Shot 2022-03-23 at 08 32 39" src="https://user-images.githubusercontent.com/3732357/159579757-3b06426f-01e8-4a02-bb2d-e7744c59918e.png">

The problem faced is that each of the RGB and white LEDs have different minimum drive currents required before they illuminate, bounding the useful PWM duty cycle range.

I also expect their forward current/brightness curve is also somewhat non-linear, and measurement/LUT would be required to get truly linear brightness control. In lieu of a data sheet, and not currently having the means to measure their light output, I'm going to assume a sufficiently linear brightness range within a defined duty cycle and drive current range.

Where I'm coming unstuck is that I can't see a way to define both a minimum and a maximum PWM duty cycle for each PWM channel independently.

`PwmRange()` appears to be most relevant here, but it only sets a maximum value, globally.

Am I missing something here? 

Happy to do some work on a PR here, but @s-hadinger given you're on most of the recent commits, do you see any issues in allowing more granular PWM duty cycle control?

**Comment by s-hadinger** [comment, marked as answer]
Hi, I'm curious what is the benefit of constant current driver for LEDs compared to actual PWM. Absence of flickering?

Technically, it's doable to have per channel min/max PMW range, the most annoying part is where to store and persist this information, and adding the additional commands to control it.

Do you by chance you ESP32 based devices?

**Reply by sfromis** [reply]
Assuming that it is Mi Bedside Lamp 2 (MJCTD02YL), that is indeed ESP32
https://github.com/mmakaay/esphome-xiaomi_bslamp2/blob/dev/doc/technical_details.md

(Since the old version MJCTD01YL has Bluetooth control, that may also be ESP32)

**Reply by barbudor** [reply]
@s-hadinger if by constant current driver you mean controlling the brightness by contolling the current then this only has disadvantages for me starting with more heat and (even) less linear control. Flickering should not be a problem as normally high frequencies are used.
Note that leds are always powered through a current driver, never voltage driver.

**Reply by s-hadinger** [reply]
Thanks, this was my understanding. With a default frequency of 977Hz, flickering is invisible. I see more advantages to adjusting brightness by duty-cycle (or constant current) rather than variable current.

**Reply by sfromis** [reply]
I'm wondering if this lamp works in a similar way to other ESP32 based Yeelight / Xiaomi Mi lamps. To get to a reasonable brightness range for both "regular lighting" up to max, and very gradually at really low brightness levels, like for nighttime lighting, several of those lamps have an additional PWM, "NL" for night light. 

Without using this, the brightness control works rather badly at lowest levels, best case one step from off to "a bit too much for night light". 

While the pinout in the description I linked have no such pin, it is certainly possible that the one doing that analysis was not aware of this detail.

**Reply by ArrEssJay** [reply]
Just to confirm @s-hadinger, yes, this is the same lamp as mentioned above. It's a single core ESP32. I've no insight into why the designers made the choice of a CC driver over PWM in this case.

I found a full range of brightness control was achievable, in that I'm able to make the LEDs illuminate barely noticeably in a pitch black room. At full duty cycle, it's bright.

I'll double check the board layout to see if there is indeed an additional PWM output I'm not aware of.

There is a small amount of flickering noticeable at extremely low illumination levels. I suspect this is a power supply regulation issue.

I have _mostly_ ported the I2C front panel driver to Berry, bar a few bugs I need to work through. Ignoring the IRQ pin and simply polling for new front panel events seems reliable. Happy to submit this for a PR once that's done.

**Reply by sfromis** [reply]
Flickering at very low brightness level is an issue also seen with other lamps, and in case of a NL pin existing, the range of that avoids weak light being at the low end of PWM, providing for flicker-free light also here.

**Reply by ArrEssJay** [reply]
Double checked and there's no separate nightlight PWM on this device. The writeup linked above is correct.

Also I'm only seeing the flickering effect on the white channel. At the brightness level impacted the light is barely visible. I suspect the flickering is always present, but not noticeable at higher brightness levels due to perceptual masking. If I get time I'll put a meter/scope on the driver output to check.

I do see some subtle colour shifts when fading on/off, suggesting there is a somewhat non-linear brightness/current curve in play, which will be difficult to account for without characterising the LEDs by measurement.

In any case, defining a usable PWM range for each LED seems like the next step.

**Reply by sfromis** [reply]
If what you need is individual control for 4 PWM channels, that could be done with Berry, as you can set duty/phase for each channel using `gpio.set_pwm`. That allows 100% control, if willing to control red/green/blue/white separately. 

If you want to stick the Tasmota features for controlling a "normal" PWM lamp, one way out could be to config PWM1-4 to unused gpio pins, and have code to catch the changes there, to replicate them to the "real" PWM channels, configured on PWM6-9, after appropriate adjustment.

The reason to skip PWM5 is that Tasmota treats PWM1-5 as a light, and PWM6 and up is for direct control (when it is ESP32).

**Reply by ArrEssJay** [reply]
I hadn't considered implementing it in Berry, at least for a proof of concept, but that might be a good route to take here given it's quite a specific use case.

**Comment by ArrEssJay** [comment]
The code below is getting there in terms of desired behaviour:

```
def setPWMOutput()
    
    var rgbw = light.get()['channels']

    for i:0..3
      # Scale 8->10 bit 
      rgbw[i] = tasmota.scale_uint(rgbw[i], 0, 255, 0, 1023)
      
      # Apply gamma
      rgbw[i] = light.gamma10(rgbw[i])
      
      # Scale within effective range for each channel
      # PWM minimums RGB: W: 61 (Noticable brightness pulsing below)
      var pwmOffset = i < 3 ? 6 : 61
      rgbw[i] = tasmota.scale_uint(rgbw[i], 0, 1023, pwmOffset, 1023)
      
    end
    
    
    # Set GPIO PWM duty cycle. No need to offset phase as power to LEDs due to CC driver.
    gpio.set_pwm(13, rgbw[0], 0)   #R
    gpio.set_pwm(14, rgbw[1], 0)   #G
    gpio.set_pwm(5, rgbw[2], 0)    #B
    gpio.set_pwm(12, rgbw[3], 0)   #W

  end
```

There are still a few issues to resolve:
- Minimum usable brightness of the white channel is a large step compared to subsequent increases in brightness. RGBW blending / gamma might need to be re-implemented in Berry.

  It may be most straightforward to pre-compute 10-bit RGB/HSV/CT -> RGBW PWM lookup tables for this device, rather than attempting to model it. Performing the math in floating point will also avoid aliasing issues when applying multiple rounds of integer range scaling.

- Fade mode is problematic. I'm trying to understand if there's a reason the esp32 `ledc` hardware PWM fade isn't used @s-hadinger. Any insight?
